include/cantera/kinetics/ReactionDerivativeManager.h

```diff
@@ -0,0 +1,140 @@
+#ifndef REACTION_DERIVATIVE_MANAGER_H
+#define REACTION_DERIVATIVE_MANAGER_H
+
+#include "cantera/zerodim.h"
+#include "cantera/kinetics.h"
+#include <vector>
+#include <unordered_map>
+#include <utility>
+
+namespace Cantera
+{
+
+class ReactionDerivative
+{
+protected:
```

Comment:
@speth  Thanks for the feedback, I have corrected most of the simple things locally and will push them soon to the PR and resolve those conversations when I do. I have some thoughts about the larger concerns as well.

> First, from a code structure standpoint, I think right now the `AdaptivePreconditioner` class is doing too much, and knows too much about specific `Reactor` types. I suspect it would be a lot cleaner if calculating Jacobian components for a reactor was done by a method on that `Reactor` class. The `ReactorNet` class would then be responsible for assembling a system Jacobian from the individual reactor Jacobians, and the `AdaptivePreconditioner` class would just be responsible for forming the preconditioner matrix from the Jacobian and serving as the interface for the factorization and solve steps.
> 

> I'd be very curious to know if there is a reasonably efficient way to build the system Jacobian from the bottom up, e.g. starting with having the derivatives of the species production rates computed by the `Kinetics` class, as suggested by @ischoegl. Even if we don't implement these derivatives that way as part of this PR, I think we should at least define what the interface should look like, i.e. what is the appropriate data structure that would let you do these evaluations without having to copy and reshuffle the data around later. I don't know what that is, and it's hard to tell what's happening now given the complete lack of documentation for the `ReactionDerivative` and `ReactionDerivativeManager` classes.
> 

I agree with this. When I started this implementation I was trying to be minimally invasive to the code base and I did not intend to create new reactor classes but that is that way that it had to work out. I suspect that it would make the most sense to have the kinetics objects apart of each reactor calculate the derivatives and write them to a given buffer which could work similarly to the integration, e.g., call `ReactorNetwork::preconditionerSetup`->`IdealGasMoleReactor::reactorPreconditionerSetup`->`Kinetics::getDerivatives`. This could also move a lot of reactor specific functionality to the respective classes. Which portion of the derivatives calculated would have to be configurable because this work intends to neglect thirdbody contributions and pressure dependence. Also, remapping of indices to write is important. These are some of the nuances which will have to be figured out. I think #1089 currently relies on the `Kinetics` object and is configurable. My implementation of `ReactionDerivativeManager` is serves the purpose of calculating the derivatives now by being initialized from the `Kinetics` objects. It is definitely hard to review code without documentation so I will get on that and consider how the functionality can be migrated to the kinetics object.

> Second, I'm a little concerned about the performance of the method right now. I tested it with a 430-species mechanism (LLNL butanol mech) and was surprised to find that adjusting the value of the "threshold" (which is what supposedly makes this an "adaptive" preconditioner) doesn't seem to have any consistent effect on the integrator's performance. I ran an ignition problem to a fixed end time for a wide range of values, and this is what I found:
>

> I'm assuming here that the variable you call the "threshold" is what McNenly et al call "eta". They suggest a default value of 1e-3, and say that this value is within 20% of the optimum, and describe its selection as a tradeoff between a small threshold that leads to a more dense preconditioner which will take longer to factorize, versus a high threshold which results in more solver iterations from a more sparse but less accurate preconditioner. Here, though, there seems to be a lot of variation in solver performance, but it's seemingly at random as the threshold changes.
> 
> so in the best case, the preconditioned iterative method is ~5 times faster, but with some threshold values can be ~3 times worse. This is compared what I would estimate to be about a factor of ~20 performance improvement for their implementation at this mechanism size.

@kyleniemeyer  and I have had some discussions about this. I am currently running some studies on a range of mechanisms and will know more about this soon. However, we were considering that the gain may be different because were are using mole fraction based reactors in comparison to LNLL's mass fraction based ones. Also, yes "threshold" is "eta".